// rheostat/lib/DefaultProgressBar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _object = require('object.assign');

var _object2 = _interopRequireDefault(_object);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _airbnbPropTypes = require('airbnb-prop-types');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactWithStyles = require('react-with-styles');

var _SliderConstants = require('./constants/SliderConstants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = (0, _airbnbPropTypes.forbidExtraProps)((0, _object2['default'])({}, _reactWithStyles.withStylesPropTypes, {
  orientation: _propTypes2['default'].string,
  disabled: _propTypes2['default'].bool,
  style: _propTypes2['default'].object
}));

var defaultProps = {
  orientation: _SliderConstants.HORIZONTAL,
  disabled: false,
  style: {}
};

function DefaultProgressBar(_ref) {
  var css = _ref.css,
      styles = _ref.styles,
      theme = _ref.theme,
      orientation = _ref.orientation,
      disabled = _ref.disabled,
      passProps = _objectWithoutProperties(_ref, ['css', 'styles', 'theme', 'orientation', 'disabled']);

  return _react2['default'].createElement('div', _extends({}, css.apply(undefined, [styles.DefaultProgressBar_progressBar].concat(_toConsumableArray(orientation === _SliderConstants.VERTICAL ? [styles.DefaultProgressBar__vertical, styles.DefaultProgressBar_background__vertical, styles.DefaultProgressBar_progressBar__vertical] : [styles.DefaultProgressBar_background__horizontal]), [disabled && styles.progressBar_disabled])), passProps));
}
DefaultProgressBar.propTypes = propTypes;
DefaultProgressBar.defaultProps = defaultProps;

exports['default'] = (0, _reactWithStyles.withStyles)(function (_ref2) {
  var _ref2$rheostat = _ref2.rheostat,
      color = _ref2$rheostat.color,
      unit = _ref2$rheostat.unit;
  return {
    DefaultProgressBar__vertical: {
      width: 3 * unit,
      height: '100%'
    },

    DefaultProgressBar_progressBar: {
      backgroundColor: color.progressBar,
      position: 'absolute'
    },

    DefaultProgressBar_progressBar__vertical: {
      height: '100%',
      width: 3 * unit
    },

    DefaultProgressBar_background__vertical: {
      height: '100%',
      top: 0,
      width: 2 * unit - 1
    },

    DefaultProgressBar_background__horizontal: {
      height: 2 * unit - 3,
      top: 0
    }
  };
})(DefaultProgressBar);
