// rheostat/lib/DefaultHandle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propTypes = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _object = require('object.assign');

var _object2 = _interopRequireDefault(_object);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _airbnbPropTypes = require('airbnb-prop-types');

var _reactWithStyles = require('react-with-styles');

var _SliderConstants = require('./constants/SliderConstants');

var _HandlePropTypes = require('./propTypes/HandlePropTypes');

var _HandlePropTypes2 = _interopRequireDefault(_HandlePropTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = exports.propTypes = (0, _airbnbPropTypes.forbidExtraProps)((0, _object2['default'])({}, _reactWithStyles.withStylesPropTypes, _HandlePropTypes2['default'], {
  'aria-valuetext': _propTypes2['default'].string,
  'aria-label': _propTypes2['default'].string
}));

var defaultProps = (0, _object2['default'])({}, _HandlePropTypes.handleDefaultProps, {
  'aria-valuetext': undefined,
  'aria-label': undefined
});

function DefaultHandle(_ref) {
  var css = _ref.css,
      styles = _ref.styles,
      orientation = _ref.orientation,
      disabled = _ref.disabled,
      handleRef = _ref.handleRef,
      theme = _ref.theme,
      passProps = _objectWithoutProperties(_ref, ['css', 'styles', 'orientation', 'disabled', 'handleRef', 'theme']);

  return _react2['default'].createElement('button', _extends({
    type: 'button',
    ref: handleRef
  }, css(styles.DefaultHandle_handle, orientation === _SliderConstants.VERTICAL ? styles.DefaultHandle_handle__vertical : styles.DefaultHandle_handle__horizontal, disabled && styles.DefaultHandle_handle__disabled), passProps));
}
DefaultHandle.propTypes = propTypes;

DefaultHandle.defaultProps = defaultProps;

exports['default'] = (0, _reactWithStyles.withStyles)(function (_ref2) {
  var _ref2$rheostat = _ref2.rheostat,
      color = _ref2$rheostat.color,
      unit = _ref2$rheostat.unit,
      constants = _ref2$rheostat.constants;
  return {
    DefaultHandle_handle: {
      width: 2 * constants.DEFAULT_HANDLE_WIDTH * unit,
      height: 2 * constants.DEFAULT_HANDLE_WIDTH * unit,
      borderWidth: 1,
      borderStyle: 'solid',
      borderColor: color.grey,
      backgroundColor: color.white,
      borderRadius: '20%',
      outline: 'none',
      zIndex: 2,
      boxShadow: '0 ' + unit / 4 + 'px ' + unit / 4 + 'px ' + String(color.textDisabled),
      ':focus': {
        boxShadow: String(color.focus) + ' 0 0 1px 1px'
      },

      ':after': {
        content: '""',
        display: 'block',
        position: 'absolute',
        backgroundColor: '#dadfe8'
      },

      ':before': {
        content: '""',
        display: 'block',
        position: 'absolute',
        backgroundColor: '#dadfe8'
      }
    },

    DefaultHandle_handle__horizontal: {
      marginLeft: -12,
      top: -5,
      ':before': {
        top: 7,
        height: 10,
        width: 1,
        left: 10
      },

      ':after': {
        top: 7,
        height: 10,
        width: 1,
        left: 13
      }
    },

    DefaultHandle_handle__vertical: {
      marginTop: -constants.DEFAULT_HANDLE_WIDTH * unit,
      left: (constants.BACKGROUND_HEIGHT - constants.DEFAULT_HANDLE_WIDTH) * unit,

      ':before': {
        top: 10
      },

      ':after': {
        top: 13,
        left: 8,
        height: 1,
        width: 10
      }
    },

    DefaultHandle_handle__disabled: {
      borderColor: color.buttons.defaultDisabledColor
    }
  };
})(DefaultHandle);
