// rheostat/lib/Slider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _object = require('object.assign');

var _object2 = _interopRequireDefault(_object);

var _reactWithStyles = require('react-with-styles');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _airbnbPropTypes = require('airbnb-prop-types');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _linear = require('./algorithms/linear');

var _linear2 = _interopRequireDefault(_linear);

var _DefaultHandle = require('./DefaultHandle');

var _DefaultHandle2 = _interopRequireDefault(_DefaultHandle);

var _DefaultProgressBar = require('./DefaultProgressBar');

var _DefaultProgressBar2 = _interopRequireDefault(_DefaultProgressBar);

var _DefaultBackground = require('./DefaultBackground');

var _DefaultBackground2 = _interopRequireDefault(_DefaultBackground);

var _OrientationPropType = require('./propTypes/OrientationPropType');

var _OrientationPropType2 = _interopRequireDefault(_OrientationPropType);

var _SliderConstants = require('./constants/SliderConstants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var has = Object.prototype.hasOwnProperty;

var PropTypeArrOfNumber = _propTypes2['default'].arrayOf(_propTypes2['default'].number);
var PropTypeReactComponent = _propTypes2['default'].oneOfType([_propTypes2['default'].func, _propTypes2['default'].string]);

function getHandleFor(ev) {
  return Number(ev.currentTarget.getAttribute('data-handle-key'));
}

function killEvent(ev) {
  ev.stopPropagation();
  ev.preventDefault();
}

var propTypes = (0, _airbnbPropTypes.forbidExtraProps)((0, _object2['default'])({}, _reactWithStyles.withStylesPropTypes, {

  // Automatically adds a top position for large when enabled
  autoAdjustVerticalPosition: _propTypes2['default'].bool,

  // the algorithm to use
  algorithm: _propTypes2['default'].shape({
    getValue: _propTypes2['default'].func,
    getPosition: _propTypes2['default'].func
  }),

  background: PropTypeReactComponent,

  // any children you pass in
  children: _propTypes2['default'].node,

  // prevent the slider from moving when clicked
  disabled: _propTypes2['default'].bool,

  // a custom handle you can pass in
  handle: PropTypeReactComponent,

  // the maximum possible value
  max: _propTypes2['default'].number,

  // the minimum possible value
  min: _propTypes2['default'].number,

  // called on click
  onClick: _propTypes2['default'].func,

  // called whenever the user is done changing values on the slider
  onChange: _propTypes2['default'].func,

  // called on key press
  onKeyPress: _propTypes2['default'].func,

  // called when you finish dragging a handle
  onSliderDragEnd: _propTypes2['default'].func,

  // called every time the slider is dragged and the value changes
  onSliderDragMove: _propTypes2['default'].func,

  // called when you start dragging a handle
  onSliderDragStart: _propTypes2['default'].func,

  // called whenever the user is actively changing the values on the slider
  // (dragging, clicked, keypress)
  onValuesUpdated: _propTypes2['default'].func,

  // the orientation
  orientation: _OrientationPropType2['default'],

  // a component for rendering the pits
  pitComponent: PropTypeReactComponent,

  // the points that pits are rendered on
  pitPoints: PropTypeArrOfNumber,

  // a custom progress bar you can pass in
  progressBar: PropTypeReactComponent,

  // should we snap?
  snap: _propTypes2['default'].bool,
  // the points we should snap to
  snapPoints: PropTypeArrOfNumber,
  // whether a proposed update is valid
  getNextHandlePosition: _propTypes2['default'].func,

  // the values
  values: PropTypeArrOfNumber
}));

var defaultProps = {
  autoAdjustVerticalPosition: false,
  children: null,
  algorithm: _linear2['default'],
  disabled: false,
  getNextHandlePosition: null,
  max: _SliderConstants.PERCENT_FULL,
  min: _SliderConstants.PERCENT_EMPTY,
  onClick: null,
  onChange: null,
  onKeyPress: null,
  onSliderDragEnd: null,
  onSliderDragMove: null,
  onSliderDragStart: null,
  onValuesUpdated: null,
  orientation: _SliderConstants.HORIZONTAL,
  pitComponent: null,
  pitPoints: [],
  snap: false,
  snapPoints: [],
  background: _DefaultBackground2['default'],
  handle: _DefaultHandle2['default'],
  progressBar: _DefaultProgressBar2['default'],
  values: [_SliderConstants.PERCENT_EMPTY]
};

var Rheostat = function (_React$Component) {
  _inherits(Rheostat, _React$Component);

  function Rheostat(props) {
    _classCallCheck(this, Rheostat);

    var _this = _possibleConstructorReturn(this, (Rheostat.__proto__ || Object.getPrototypeOf(Rheostat)).call(this, props));

    var _this$props = _this.props,
        algorithm = _this$props.algorithm,
        max = _this$props.max,
        min = _this$props.min,
        values = _this$props.values;


    _this.state = {
      handlePos: values.map(function (value) {
        return algorithm.getPosition(value, min, max);
      }),
      handleDimensions: 0,
      slidingIndex: null,
      values: values
    };

    _this.getPublicState = _this.getPublicState.bind(_this);
    _this.getSliderBoundingBox = _this.getSliderBoundingBox.bind(_this);
    _this.getProgressStyle = _this.getProgressStyle.bind(_this);
    _this.getMinValue = _this.getMinValue.bind(_this);
    _this.getMaxValue = _this.getMaxValue.bind(_this);
    _this.getHandleDimensions = _this.getHandleDimensions.bind(_this);
    _this.getClosestSnapPoint = _this.getClosestSnapPoint.bind(_this);
    _this.getSnapPosition = _this.getSnapPosition.bind(_this);
    _this.getNextPositionForKey = _this.getNextPositionForKey.bind(_this);
    _this.getNextState = _this.getNextState.bind(_this);
    _this.handleClick = _this.handleClick.bind(_this);
    _this.getClosestHandle = _this.getClosestHandle.bind(_this);
    _this.setStartSlide = _this.setStartSlide.bind(_this);
    _this.startMouseSlide = _this.startMouseSlide.bind(_this);
    _this.startTouchSlide = _this.startTouchSlide.bind(_this);
    _this.handleMouseSlide = _this.handleMouseSlide.bind(_this);
    _this.handleTouchSlide = _this.handleTouchSlide.bind(_this);
    _this.handleSlide = _this.handleSlide.bind(_this);
    _this.endSlide = _this.endSlide.bind(_this);
    _this.handleKeydown = _this.handleKeydown.bind(_this);
    _this.validatePosition = _this.validatePosition.bind(_this);
    _this.validateValues = _this.validateValues.bind(_this);
    _this.canMove = _this.canMove.bind(_this);
    _this.fireChangeEvent = _this.fireChangeEvent.bind(_this);
    _this.slideTo = _this.slideTo.bind(_this);
    _this.updateNewValues = _this.updateNewValues.bind(_this);
    _this.setRef = _this.setRef.bind(_this);
    _this.setHandleNode = _this.setHandleNode.bind(_this);
    _this.setHandleContainerNode = _this.setHandleContainerNode.bind(_this);
    _this.positionPercent = _this.positionPercent.bind(_this);
    _this.invalidatePitStyleCache = _this.invalidatePitStyleCache.bind(_this);

    _this.pitStyleCache = {};
    return _this;
  }

  _createClass(Rheostat, [{
    key: 'componentDidMount',
    value: function () {
      function componentDidMount() {
        var _this2 = this;

        // Note: This occurs in a timeout because styles need to be applied first
        this.handleDimensionsTimeout = setTimeout(function () {
          _this2.handleDimensionsTimeout = null;
          _this2.setState({ handleDimensions: _this2.getHandleDimensions() });
        }, 0);
      }

      return componentDidMount;
    }()
  }, {
    key: 'componentWillReceiveProps',
    value: function () {
      function componentWillReceiveProps(nextProps) {
        var _props = this.props,
            disabled = _props.disabled,
            min = _props.min,
            max = _props.max,
            orientation = _props.orientation,
            pitPoints = _props.pitPoints,
            algorithm = _props.algorithm;
        var _state = this.state,
            values = _state.values,
            slidingIndex = _state.slidingIndex;


        var minMaxChanged = nextProps.min !== min || nextProps.max !== max;

        var valuesChanged = values.length !== nextProps.values.length || values.some(function (value, idx) {
          return nextProps.values[idx] !== value;
        });

        var orientationChanged = nextProps.orientation !== orientation;

        var algorithmChanged = nextProps.algorithm !== algorithm;

        var pitPointsChanged = nextProps.pitPoints !== pitPoints;

        var willBeDisabled = nextProps.disabled && !disabled;

        if (minMaxChanged || valuesChanged) this.updateNewValues(nextProps);

        if (willBeDisabled && slidingIndex !== null) {
          this.endSlide();
        }

        if (minMaxChanged || pitPointsChanged || orientationChanged || algorithmChanged) {
          this.invalidatePitStyleCache();
        }
      }

      return componentWillReceiveProps;
    }()
  }, {
    key: 'componentWillUnmount',
    value: function () {
      function componentWillUnmount() {
        if (this.handleDimensionsTimeout) {
          clearTimeout(this.handleDimensionsTimeout);
          this.handleDimensionsTimeout = null;
        }
      }

      return componentWillUnmount;
    }()
  }, {
    key: 'getPublicState',
    value: function () {
      function getPublicState() {
        var values = this.state.values;
        var _props2 = this.props,
            min = _props2.min,
            max = _props2.max;


        return {
          max: max,
          min: min,
          values: values
        };
      }

      return getPublicState;
    }()
  }, {
    key: 'getSliderBoundingBox',
    value: function () {
      function getSliderBoundingBox() {
        var rect = this.handleContainerNode.getBoundingClientRect();
        return {
          height: rect.height || this.handleContainerNode.clientHeight,
          left: rect.left,
          right: rect.right,
          top: rect.top,
          width: rect.width || this.handleContainerNode.clientWidth
        };
      }

      return getSliderBoundingBox;
    }()
  }, {
    key: 'getProgressStyle',
    value: function () {
      function getProgressStyle(idx) {
        var orientation = this.props.orientation;
        var handlePos = this.state.handlePos;


        var value = handlePos[idx];

        if (idx === 0) {
          return orientation === _SliderConstants.VERTICAL ? { height: String(value) + '%', top: 0 } : { left: 0, width: String(value) + '%' };
        }

        var prevValue = handlePos[idx - 1];
        var diffValue = value - prevValue;

        return orientation === _SliderConstants.VERTICAL ? { height: diffValue + '%', top: String(prevValue) + '%' } : { left: String(prevValue) + '%', width: diffValue + '%' };
      }

      return getProgressStyle;
    }()
  }, {
    key: 'getMinValue',
    value: function () {
      function getMinValue(idx) {
        var min = this.props.min;
        var values = this.state.values;


        return values[idx - 1] ? Math.max(min, values[idx - 1]) : min;
      }

      return getMinValue;
    }()
  }, {
    key: 'getMaxValue',
    value: function () {
      function getMaxValue(idx) {
        var max = this.props.max;
        var values = this.state.values;


        return values[idx + 1] ? Math.min(max, values[idx + 1]) : max;
      }

      return getMaxValue;
    }()
  }, {
    key: 'getClosestSnapPoint',
    value: function () {
      function getClosestSnapPoint(value) {
        var snapPoints = this.props.snapPoints;

        if (!snapPoints.length) return value;

        return snapPoints.reduce(function (snapTo, snap) {
          return Math.abs(snapTo - value) < Math.abs(snap - value) ? snapTo : snap;
        });
      }

      return getClosestSnapPoint;
    }()
  }, {
    key: 'getHandleDimensions',
    value: function () {
      function getHandleDimensions() {
        var orientation = this.props.orientation;

        if (!this.handleNode) return 0;

        return orientation === _SliderConstants.VERTICAL ? this.handleNode.clientHeight : this.handleNode.clientWidth;
      }

      return getHandleDimensions;
    }()
  }, {
    key: 'getSnapPosition',
    value: function () {
      function getSnapPosition(positionPercent) {
        var _props3 = this.props,
            algorithm = _props3.algorithm,
            max = _props3.max,
            min = _props3.min,
            snap = _props3.snap;


        if (!snap) return positionPercent;
        var value = algorithm.getValue(positionPercent, min, max);
        var snapValue = this.getClosestSnapPoint(value);
        return algorithm.getPosition(snapValue, min, max);
      }

      return getSnapPosition;
    }()
  }, {
    key: 'getNextPositionForKey',
    value: function () {
      function getNextPositionForKey(idx, keyCode) {
        var _stepMultiplier;

        var _state2 = this.state,
            handlePos = _state2.handlePos,
            values = _state2.values;
        var _props4 = this.props,
            algorithm = _props4.algorithm,
            max = _props4.max,
            min = _props4.min,
            snapPoints = _props4.snapPoints,
            shouldSnap = _props4.snap;


        var proposedValue = values[idx];
        var proposedPercentage = handlePos[idx];
        var originalPercentage = proposedPercentage;
        var stepValue = 1;

        if (max >= 100) {
          proposedPercentage = Math.round(proposedPercentage);
        } else {
          stepValue = 100 / (max - min);
        }

        var currentIndex = null;

        if (shouldSnap) {
          currentIndex = snapPoints.indexOf(this.getClosestSnapPoint(values[idx]));
        }

        var stepMultiplier = (_stepMultiplier = {}, _defineProperty(_stepMultiplier, _SliderConstants.KEYS.LEFT, function (v) {
          return v * -1;
        }), _defineProperty(_stepMultiplier, _SliderConstants.KEYS.RIGHT, function (v) {
          return v * 1;
        }), _defineProperty(_stepMultiplier, _SliderConstants.KEYS.UP, function (v) {
          return v * 1;
        }), _defineProperty(_stepMultiplier, _SliderConstants.KEYS.DOWN, function (v) {
          return v * -1;
        }), _defineProperty(_stepMultiplier, _SliderConstants.KEYS.PAGE_DOWN, function (v) {
          return v > 1 ? -v : v * -10;
        }), _defineProperty(_stepMultiplier, _SliderConstants.KEYS.PAGE_UP, function (v) {
          return v > 1 ? v : v * 10;
        }), _stepMultiplier);

        if (has.call(stepMultiplier, keyCode)) {
          proposedPercentage += stepMultiplier[keyCode](stepValue);

          if (shouldSnap) {
            if (proposedPercentage > originalPercentage) {
              // move cursor right unless overflow
              if (currentIndex < snapPoints.length - 1) {
                proposedValue = snapPoints[currentIndex + 1];
              }
              // move cursor left unless there is overflow
            } else if (currentIndex > 0) {
              proposedValue = snapPoints[currentIndex - 1];
            }
          }
        } else if (keyCode === _SliderConstants.KEYS.HOME) {
          proposedPercentage = _SliderConstants.PERCENT_EMPTY;

          if (shouldSnap) {
            var _snapPoints = _slicedToArray(snapPoints, 1);

            proposedValue = _snapPoints[0];
          }
        } else if (keyCode === _SliderConstants.KEYS.END) {
          proposedPercentage = _SliderConstants.PERCENT_FULL;

          if (shouldSnap) {
            proposedValue = snapPoints[snapPoints.length - 1];
          }
        } else {
          return null;
        }

        return shouldSnap ? algorithm.getPosition(proposedValue, min, max) : proposedPercentage;
      }

      return getNextPositionForKey;
    }()
  }, {
    key: 'getNextState',
    value: function () {
      function getNextState(idx, proposedPosition) {
        var handlePos = this.state.handlePos;
        var _props5 = this.props,
            max = _props5.max,
            min = _props5.min,
            algorithm = _props5.algorithm;


        var actualPosition = this.validatePosition(idx, proposedPosition);

        var nextHandlePos = handlePos.map(function (pos, index) {
          return index === idx ? actualPosition : pos;
        });

        return {
          handlePos: nextHandlePos,
          values: nextHandlePos.map(function (pos) {
            return algorithm.getValue(pos, min, max);
          })
        };
      }

      return getNextState;
    }()
  }, {
    key: 'getClosestHandle',
    value: function () {
      function getClosestHandle(positionPercent) {
        var handlePos = this.state.handlePos;


        return handlePos.reduce(function (closestIdx, node, idx) {
          var challenger = Math.abs(handlePos[idx] - positionPercent);
          var current = Math.abs(handlePos[closestIdx] - positionPercent);
          return challenger < current ? idx : closestIdx;
        }, 0);
      }

      return getClosestHandle;
    }()
  }, {
    key: 'setHandleNode',
    value: function () {
      function setHandleNode(node) {
        this.handleNode = node;
      }

      return setHandleNode;
    }()
  }, {
    key: 'setHandleContainerNode',
    value: function () {
      function setHandleContainerNode(node) {
        this.handleContainerNode = node;
      }

      return setHandleContainerNode;
    }()
  }, {
    key: 'setStartSlide',
    value: function () {
      function setStartSlide(ev) {
        var sliderBox = this.getSliderBoundingBox();
        this.setState({
          handleDimensions: this.getHandleDimensions(ev, sliderBox),
          slidingIndex: getHandleFor(ev)
        });
      }

      return setStartSlide;
    }()
  }, {
    key: 'setRef',
    value: function () {
      function setRef(ref) {
        this.rheostat = ref;
      }

      return setRef;
    }()
  }, {
    key: 'startMouseSlide',
    value: function () {
      function startMouseSlide(ev) {
        var onSliderDragStart = this.props.onSliderDragStart;


        this.setStartSlide(ev, ev.clientX, ev.clientY);

        if (typeof document.addEventListener === 'function') {
          document.addEventListener('mousemove', this.handleMouseSlide, false);
          document.addEventListener('mouseup', this.endSlide, false);
        } else {
          document.attachEvent('onmousemove', this.handleMouseSlide);
          document.attachEvent('onmouseup', this.endSlide);
        }

        if (onSliderDragStart) onSliderDragStart();

        killEvent(ev);
      }

      return startMouseSlide;
    }()
  }, {
    key: 'startTouchSlide',
    value: function () {
      function startTouchSlide(ev) {
        var onSliderDragStart = this.props.onSliderDragStart;


        if (ev.changedTouches.length > 1) return;

        var touch = ev.changedTouches[0];

        this.setStartSlide(ev, touch.clientX, touch.clientY);

        document.addEventListener('touchmove', this.handleTouchSlide, false);
        document.addEventListener('touchend', this.endSlide, false);

        if (onSliderDragStart) onSliderDragStart();

        killEvent(ev);
      }

      return startTouchSlide;
    }()
  }, {
    key: 'handleMouseSlide',
    value: function () {
      function handleMouseSlide(ev) {
        var slidingIndex = this.state.slidingIndex;


        if (slidingIndex === null) return;
        this.handleSlide(ev.clientX, ev.clientY);
        killEvent(ev);
      }

      return handleMouseSlide;
    }()
  }, {
    key: 'handleTouchSlide',
    value: function () {
      function handleTouchSlide(ev) {
        var slidingIndex = this.state.slidingIndex;


        if (slidingIndex === null) return;

        if (ev.changedTouches.length > 1) {
          this.endSlide();
          return;
        }

        var touch = ev.changedTouches[0];

        this.handleSlide(touch.clientX, touch.clientY);
        killEvent(ev);
      }

      return handleTouchSlide;
    }()
  }, {
    key: 'positionPercent',
    value: function () {
      function positionPercent(x, y, sliderBox) {
        var orientation = this.props.orientation;

        if (orientation === _SliderConstants.VERTICAL) {
          return (y - sliderBox.top) / sliderBox.height * _SliderConstants.PERCENT_FULL;
        }
        return (x - sliderBox.left) / sliderBox.width * _SliderConstants.PERCENT_FULL;
      }

      return positionPercent;
    }()
  }, {
    key: 'handleSlide',
    value: function () {
      function handleSlide(x, y) {
        var onSliderDragMove = this.props.onSliderDragMove;
        var idx = this.state.slidingIndex;

        var sliderBox = this.getSliderBoundingBox();
        var positionPercent = this.positionPercent(x, y, sliderBox);

        this.slideTo(idx, positionPercent);

        if (this.canMove(idx, positionPercent)) {
          if (onSliderDragMove) onSliderDragMove();
        }
      }

      return handleSlide;
    }()
  }, {
    key: 'endSlide',
    value: function () {
      function endSlide() {
        var _this3 = this;

        var _props6 = this.props,
            onSliderDragEnd = _props6.onSliderDragEnd,
            snap = _props6.snap;
        var _state3 = this.state,
            slidingIndex = _state3.slidingIndex,
            handlePos = _state3.handlePos;


        this.setState({ slidingIndex: null });

        if (typeof document.removeEventListener === 'function') {
          document.removeEventListener('mouseup', this.endSlide, false);
          document.removeEventListener('touchend', this.endSlide, false);
          document.removeEventListener('touchmove', this.handleTouchSlide, false);
          document.removeEventListener('mousemove', this.handleMouseSlide, false);
        } else {
          document.detachEvent('onmousemove', this.handleMouseSlide);
          document.detachEvent('onmouseup', this.endSlide);
        }

        if (onSliderDragEnd) onSliderDragEnd();
        if (snap) {
          var positionPercent = this.getSnapPosition(handlePos[slidingIndex]);
          this.slideTo(slidingIndex, positionPercent, function () {
            return _this3.fireChangeEvent();
          });
        } else {
          this.fireChangeEvent();
        }
      }

      return endSlide;
    }()
  }, {
    key: 'handleClick',
    value: function () {
      function handleClick(ev) {
        var _this4 = this;

        if (ev.target.getAttribute('data-handle-key')) {
          return;
        }

        var _props7 = this.props,
            onClick = _props7.onClick,
            orientation = _props7.orientation;

        // Calculate the position of the slider on the page so we can determine
        // the position where you click in relativity.

        var sliderBox = this.getSliderBoundingBox();

        var positionDecimal = orientation === _SliderConstants.VERTICAL ? (ev.clientY - sliderBox.top) / sliderBox.height : (ev.clientX - sliderBox.left) / sliderBox.width;

        var positionPercent = positionDecimal * _SliderConstants.PERCENT_FULL;

        var handleId = this.getClosestHandle(positionPercent);

        var validPositionPercent = this.getSnapPosition(positionPercent);

        // Move the handle there
        this.slideTo(handleId, validPositionPercent, function () {
          return _this4.fireChangeEvent();
        });

        if (onClick) onClick();
      }

      return handleClick;
    }()
  }, {
    key: 'handleKeydown',
    value: function () {
      function handleKeydown(ev) {
        var _this5 = this;

        var onKeyPress = this.props.onKeyPress;

        var idx = getHandleFor(ev);

        if (ev.keyCode === _SliderConstants.KEYS.ESC) {
          ev.currentTarget.blur();
          return;
        }

        var proposedPercentage = this.getNextPositionForKey(idx, ev.keyCode);

        if (proposedPercentage === null) return;

        if (this.canMove(idx, proposedPercentage)) {
          this.slideTo(idx, proposedPercentage, function () {
            return _this5.fireChangeEvent();
          });
          if (onKeyPress) onKeyPress();
        }

        killEvent(ev);
      }

      return handleKeydown;
    }()

    // Apply user adjustments to position

  }, {
    key: 'userAdjustPosition',
    value: function () {
      function userAdjustPosition(idx, proposedPosition) {
        var getNextHandlePosition = this.props.getNextHandlePosition;

        var nextPosition = proposedPosition;
        if (getNextHandlePosition) {
          nextPosition = parseFloat(getNextHandlePosition(idx, proposedPosition));

          if (Number.isNaN(nextPosition) || nextPosition < _SliderConstants.PERCENT_EMPTY || nextPosition > _SliderConstants.PERCENT_FULL) {
            throw new TypeError('getNextHandlePosition returned invalid position. Valid positions are floats between 0 and 100');
          }
        }

        return nextPosition;
      }

      return userAdjustPosition;
    }()

    // Make sure the proposed position respects the bounds and
    // does not collide with other handles too much.

  }, {
    key: 'validatePosition',
    value: function () {
      function validatePosition(idx, proposedPosition) {
        var _state4 = this.state,
            handlePos = _state4.handlePos,
            handleDimensions = _state4.handleDimensions;


        var nextPosition = this.userAdjustPosition(idx, proposedPosition);

        var orientation = this.props.orientation;

        var sliderBox = this.getSliderBoundingBox();

        var handlePercentage = orientation === _SliderConstants.VERTICAL ? handleDimensions / sliderBox.height * _SliderConstants.PERCENT_FULL / 2 : handleDimensions / sliderBox.width * _SliderConstants.PERCENT_FULL / 2;

        return Math.max(Math.min(nextPosition, handlePos[idx + 1] !== undefined ? handlePos[idx + 1] - handlePercentage : _SliderConstants.PERCENT_FULL // 100% is the highest value
        ), handlePos[idx - 1] !== undefined ? handlePos[idx - 1] + handlePercentage : _SliderConstants.PERCENT_EMPTY // 0% is the lowest value
        );
      }

      return validatePosition;
    }()
  }, {
    key: 'validateValues',
    value: function () {
      function validateValues(proposedValues, props) {
        var _ref = props || this.props,
            max = _ref.max,
            min = _ref.min;

        return proposedValues.map(function (value, idx, values) {
          var realValue = Math.max(Math.min(value, max), min);

          if (values.length && realValue < values[idx - 1]) {
            return values[idx - 1];
          }

          return realValue;
        });
      }

      return validateValues;
    }()

    // Can we move the slider to the given position?

  }, {
    key: 'canMove',
    value: function () {
      function canMove(idx, proposedPosition) {
        var _state5 = this.state,
            handlePos = _state5.handlePos,
            handleDimensions = _state5.handleDimensions;
        var orientation = this.props.orientation;

        var sliderBox = this.getSliderBoundingBox();

        var handlePercentage = orientation === _SliderConstants.VERTICAL ? handleDimensions / sliderBox.height * _SliderConstants.PERCENT_FULL / 2 : handleDimensions / sliderBox.width * _SliderConstants.PERCENT_FULL / 2;

        if (proposedPosition < _SliderConstants.PERCENT_EMPTY) return false;
        if (proposedPosition > _SliderConstants.PERCENT_FULL) return false;

        var nextHandlePosition = handlePos[idx + 1] !== undefined ? handlePos[idx + 1] - handlePercentage : Infinity;

        if (proposedPosition > nextHandlePosition) return false;

        var prevHandlePosition = handlePos[idx - 1] !== undefined ? handlePos[idx - 1] + handlePercentage : -Infinity;

        if (proposedPosition < prevHandlePosition) return false;

        return true;
      }

      return canMove;
    }()
  }, {
    key: 'fireChangeEvent',
    value: function () {
      function fireChangeEvent() {
        var onChange = this.props.onChange;

        if (onChange) onChange(this.getPublicState());
      }

      return fireChangeEvent;
    }()
  }, {
    key: 'slideTo',
    value: function () {
      function slideTo(idx, proposedPosition, onAfterSet) {
        var _this6 = this;

        var onValuesUpdated = this.props.onValuesUpdated;

        var nextState = this.getNextState(idx, proposedPosition);

        this.setState(nextState, function () {
          if (onValuesUpdated) onValuesUpdated(_this6.getPublicState());
          if (onAfterSet) onAfterSet();
        });
      }

      return slideTo;
    }()
  }, {
    key: 'updateNewValues',
    value: function () {
      function updateNewValues(nextProps) {
        var slidingIndex = this.state.slidingIndex;

        // Don't update while the slider is sliding

        if (slidingIndex !== null) {
          return;
        }
        var algorithm = this.props.algorithm;
        var max = nextProps.max,
            min = nextProps.min,
            values = nextProps.values;


        var nextValues = this.validateValues(values, nextProps);
        this.setState({
          handlePos: nextValues.map(function (value) {
            return algorithm.getPosition(value, min, max);
          }),
          values: nextValues
        });
      }

      return updateNewValues;
    }()
  }, {
    key: 'invalidatePitStyleCache',
    value: function () {
      function invalidatePitStyleCache() {
        this.pitStyleCache = {};
      }

      return invalidatePitStyleCache;
    }()
  }, {
    key: 'render',
    value: function () {
      function render() {
        var _this7 = this;

        var _props8 = this.props,
            css = _props8.css,
            autoAdjustVerticalPosition = _props8.autoAdjustVerticalPosition,
            algorithm = _props8.algorithm,
            children = _props8.children,
            disabled = _props8.disabled,
            Handle = _props8.handle,
            max = _props8.max,
            min = _props8.min,
            orientation = _props8.orientation,
            PitComponent = _props8.pitComponent,
            pitPoints = _props8.pitPoints,
            Background = _props8.background,
            ProgressBar = _props8.progressBar,
            styles = _props8.styles;
        var _state6 = this.state,
            handlePos = _state6.handlePos,
            values = _state6.values;


        return (
          // eslint-disable-next-line jsx-a11y/click-events-have-key-events
          _react2['default'].createElement(
            'div',
            _extends({
              onClick: disabled ? undefined : this.handleClick
            }, css(styles.rheostat, autoAdjustVerticalPosition && styles.autoAdjustVerticalPosition, orientation === _SliderConstants.VERTICAL && styles.rheostat__vertical)),
            !!Background && _react2['default'].createElement(Background, {
              orientation: orientation
            }),
            _react2['default'].createElement(
              'div',
              _extends({
                ref: this.setHandleContainerNode
              }, css(styles.handleContainer)),
              handlePos.map(function (pos, idx) {
                var handleStyle = orientation === _SliderConstants.VERTICAL ? { top: String(pos) + '%', position: 'absolute' } : { left: String(pos) + '%', position: 'absolute' };

                return _react2['default'].createElement(Handle, {
                  'aria-valuemax': _this7.getMaxValue(idx),
                  'aria-valuemin': _this7.getMinValue(idx),
                  'aria-valuenow': values[idx],
                  'aria-disabled': disabled,
                  'data-handle-key': idx,
                  key: idx /* eslint-disable-line react/no-array-index-key */,
                  orientation: orientation,
                  disabled: disabled,
                  onClick: _this7.killEvent,
                  onKeyDown: disabled ? undefined : _this7.handleKeydown,
                  onMouseDown: disabled ? undefined : _this7.startMouseSlide,
                  onTouchStart: disabled ? undefined : _this7.startTouchSlide,
                  handleRef: _this7.setHandleNode,
                  role: 'slider',
                  style: handleStyle,
                  tabIndex: 0
                });
              })
            ),
            !!ProgressBar && handlePos.map(function (node, idx, arr) {
              if (idx === 0 && arr.length > 1) {
                return null;
              }
              return _react2['default'].createElement(ProgressBar, {
                key: idx /* eslint-disable-line react/no-array-index-key */,
                style: _this7.getProgressStyle(idx),
                disabled: disabled
              });
            }),
            PitComponent && pitPoints.map(function (n) {
              var pitStyle = _this7.pitStyleCache[n];

              if (!pitStyle) {
                var pos = algorithm.getPosition(n, min, max);
                pitStyle = orientation === 'vertical' ? { top: String(pos) + '%', position: 'absolute' } : { left: String(pos) + '%', position: 'absolute' };
                _this7.pitStyleCache[n] = pitStyle;
              }

              return _react2['default'].createElement(
                PitComponent,
                { key: n, style: pitStyle },
                n
              );
            }),
            children
          )
        );
      }

      return render;
    }()
  }]);

  return Rheostat;
}(_react2['default'].Component);

Rheostat.propTypes = propTypes;
Rheostat.defaultProps = defaultProps;

exports['default'] = (0, _reactWithStyles.withStyles)(function (_ref2) {
  var _ref2$rheostat = _ref2.rheostat,
      color = _ref2$rheostat.color,
      unit = _ref2$rheostat.unit,
      responsive = _ref2$rheostat.responsive;
  return {
    rheostat: {
      position: 'relative',
      overflow: 'visible'
    },

    autoAdjustVerticalPosition: _defineProperty({}, responsive.largeAndAbove, {
      top: 1.5 * unit
    }),

    rheostat__vertical: {
      height: '100%'
    },

    handleContainer: {
      height: 2 * unit - 1,
      top: -2,
      left: -2,
      bottom: 4,
      width: '100%',
      position: 'absolute'
    },

    rheostat_background: {
      backgroundColor: color.white,
      border: '1px solid ' + String(color.grey),
      position: 'relative'
    },

    rheostat_background__horizontal: {
      height: 2 * unit - 1,
      top: -2,
      left: -2,
      bottom: 4,
      width: '100%'
    },

    rheostat_background__vertical: {
      width: 2 * unit - 1,
      top: 0,
      height: '100%'
    }
  };
})(Rheostat);
