// 1-liners/index.js
'use strict';

exports.__esModule = true;
// istanbul ignore next

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _always = require('./always');

var _always2 = _interopRequireDefault(_always);

var _and = require('./and');

var _and2 = _interopRequireDefault(_and);

var _assign = require('./assign');

var _assign2 = _interopRequireDefault(_assign);

var _average = require('./average');

var _average2 = _interopRequireDefault(_average);

var _between = require('./between');

var _between2 = _interopRequireDefault(_between);

var _bind = require('./bind');

var _bind2 = _interopRequireDefault(_bind);

var _bitAnd = require('./bitAnd');

var _bitAnd2 = _interopRequireDefault(_bitAnd);

var _bitOr = require('./bitOr');

var _bitOr2 = _interopRequireDefault(_bitOr);

var _butLast = require('./butLast');

var _butLast2 = _interopRequireDefault(_butLast);

var _by = require('./by');

var _by2 = _interopRequireDefault(_by);

var _charAt = require('./charAt');

var _charAt2 = _interopRequireDefault(_charAt);

var _charCodeAt = require('./charCodeAt');

var _charCodeAt2 = _interopRequireDefault(_charCodeAt);

var _codePointAt = require('./codePointAt');

var _codePointAt2 = _interopRequireDefault(_codePointAt);

var _compose = require('./compose');

var _compose2 = _interopRequireDefault(_compose);

var _composeAll = require('./composeAll');

var _composeAll2 = _interopRequireDefault(_composeAll);

var _concat = require('./concat');

var _concat2 = _interopRequireDefault(_concat);

var _converge = require('./converge');

var _converge2 = _interopRequireDefault(_converge);

var _curry = require('./curry');

var _curry2 = _interopRequireDefault(_curry);

var _curryRight = require('./curryRight');

var _curryRight2 = _interopRequireDefault(_curryRight);

var _dec = require('./dec');

var _dec2 = _interopRequireDefault(_dec);

var _drop = require('./drop');

var _drop2 = _interopRequireDefault(_drop);

var _endsWith = require('./endsWith');

var _endsWith2 = _interopRequireDefault(_endsWith);

var _endsWithAt = require('./endsWithAt');

var _endsWithAt2 = _interopRequireDefault(_endsWithAt);

var _equal = require('./equal');

var _equal2 = _interopRequireDefault(_equal);

var _every = require('./every');

var _every2 = _interopRequireDefault(_every);

var _exec = require('./exec');

var _exec2 = _interopRequireDefault(_exec);

var _explode = require('./explode');

var _explode2 = _interopRequireDefault(_explode);

var _filter = require('./filter');

var _filter2 = _interopRequireDefault(_filter);

var _flatMap = require('./flatMap');

var _flatMap2 = _interopRequireDefault(_flatMap);

var _flip = require('./flip');

var _flip2 = _interopRequireDefault(_flip);

var _fold = require('./fold');

var _fold2 = _interopRequireDefault(_fold);

var _foldRight = require('./foldRight');

var _foldRight2 = _interopRequireDefault(_foldRight);

var _forEach = require('./forEach');

var _forEach2 = _interopRequireDefault(_forEach);

var _greaterOrEqual = require('./greaterOrEqual');

var _greaterOrEqual2 = _interopRequireDefault(_greaterOrEqual);

var _greaterThan = require('./greaterThan');

var _greaterThan2 = _interopRequireDefault(_greaterThan);

var _hasOwnProperty = require('./hasOwnProperty');

var _hasOwnProperty2 = _interopRequireDefault(_hasOwnProperty);

var _head = require('./head');

var _head2 = _interopRequireDefault(_head);

var _identity = require('./identity');

var _identity2 = _interopRequireDefault(_identity);

var _ifThen = require('./ifThen');

var _ifThen2 = _interopRequireDefault(_ifThen);

var _ifThenElse = require('./ifThenElse');

var _ifThenElse2 = _interopRequireDefault(_ifThenElse);

var _implode = require('./implode');

var _implode2 = _interopRequireDefault(_implode);

var _inc = require('./inc');

var _inc2 = _interopRequireDefault(_inc);

var _includes = require('./includes');

var _includes2 = _interopRequireDefault(_includes);

var _indexOf = require('./indexOf');

var _indexOf2 = _interopRequireDefault(_indexOf);

var _isBetween = require('./isBetween');

var _isBetween2 = _interopRequireDefault(_isBetween);

var _isBoolean = require('./isBoolean');

var _isBoolean2 = _interopRequireDefault(_isBoolean);

var _isFalse = require('./isFalse');

var _isFalse2 = _interopRequireDefault(_isFalse);

var _isFalsy = require('./isFalsy');

var _isFalsy2 = _interopRequireDefault(_isFalsy);

var _isFunction = require('./isFunction');

var _isFunction2 = _interopRequireDefault(_isFunction);

var _isNull = require('./isNull');

var _isNull2 = _interopRequireDefault(_isNull);

var _isNumber = require('./isNumber');

var _isNumber2 = _interopRequireDefault(_isNumber);

var _isObject = require('./isObject');

var _isObject2 = _interopRequireDefault(_isObject);

var _isPlainObject = require('./isPlainObject');

var _isPlainObject2 = _interopRequireDefault(_isPlainObject);

var _isString = require('./isString');

var _isString2 = _interopRequireDefault(_isString);

var _isTrue = require('./isTrue');

var _isTrue2 = _interopRequireDefault(_isTrue);

var _isTruthy = require('./isTruthy');

var _isTruthy2 = _interopRequireDefault(_isTruthy);

var _isTypeOf = require('./isTypeOf');

var _isTypeOf2 = _interopRequireDefault(_isTypeOf);

var _isUndefined = require('./isUndefined');

var _isUndefined2 = _interopRequireDefault(_isUndefined);

var _isUnknown = require('./isUnknown');

var _isUnknown2 = _interopRequireDefault(_isUnknown);

var _join = require('./join');

var _join2 = _interopRequireDefault(_join);

var _keys = require('./keys');

var _keys2 = _interopRequireDefault(_keys);

var _last = require('./last');

var _last2 = _interopRequireDefault(_last);

var _lastIndexOf = require('./lastIndexOf');

var _lastIndexOf2 = _interopRequireDefault(_lastIndexOf);

var _length = require('./length');

var _length2 = _interopRequireDefault(_length);

var _lessOrEqual = require('./lessOrEqual');

var _lessOrEqual2 = _interopRequireDefault(_lessOrEqual);

var _lessThan = require('./lessThan');

var _lessThan2 = _interopRequireDefault(_lessThan);

var _localeCompare = require('./localeCompare');

var _localeCompare2 = _interopRequireDefault(_localeCompare);

var _looseEqual = require('./looseEqual');

var _looseEqual2 = _interopRequireDefault(_looseEqual);

var _map = require('./map');

var _map2 = _interopRequireDefault(_map);

var _match = require('./match');

var _match2 = _interopRequireDefault(_match);

var _max = require('./max');

var _max2 = _interopRequireDefault(_max);

var _method = require('./method');

var _method2 = _interopRequireDefault(_method);

var _min = require('./min');

var _min2 = _interopRequireDefault(_min);

var _minus = require('./minus');

var _minus2 = _interopRequireDefault(_minus);

var _nand = require('./nand');

var _nand2 = _interopRequireDefault(_nand);

var _noop = require('./noop');

var _noop2 = _interopRequireDefault(_noop);

var _nor = require('./nor');

var _nor2 = _interopRequireDefault(_nor);

var _normalize = require('./normalize');

var _normalize2 = _interopRequireDefault(_normalize);

var _not = require('./not');

var _not2 = _interopRequireDefault(_not);

var _nth = require('./nth');

var _nth2 = _interopRequireDefault(_nth);

var _omit = require('./omit');

var _omit2 = _interopRequireDefault(_omit);

var _or = require('./or');

var _or2 = _interopRequireDefault(_or);

var _pick = require('./pick');

var _pick2 = _interopRequireDefault(_pick);

var _pipe = require('./pipe');

var _pipe2 = _interopRequireDefault(_pipe);

var _pipeAll = require('./pipeAll');

var _pipeAll2 = _interopRequireDefault(_pipeAll);

var _plus = require('./plus');

var _plus2 = _interopRequireDefault(_plus);

var _product = require('./product');

var _product2 = _interopRequireDefault(_product);

var _property = require('./property');

var _property2 = _interopRequireDefault(_property);

var _push = require('./push');

var _push2 = _interopRequireDefault(_push);

var _put = require('./put');

var _put2 = _interopRequireDefault(_put);

var _reduce = require('./reduce');

var _reduce2 = _interopRequireDefault(_reduce);

var _reduceRight = require('./reduceRight');

var _reduceRight2 = _interopRequireDefault(_reduceRight);

var _repeat = require('./repeat');

var _repeat2 = _interopRequireDefault(_repeat);

var _replace = require('./replace');

var _replace2 = _interopRequireDefault(_replace);

var _search = require('./search');

var _search2 = _interopRequireDefault(_search);

var _shallowClone = require('./shallowClone');

var _shallowClone2 = _interopRequireDefault(_shallowClone);

var _shave = require('./shave');

var _shave2 = _interopRequireDefault(_shave);

var _signum = require('./signum');

var _signum2 = _interopRequireDefault(_signum);

var _slice = require('./slice');

var _slice2 = _interopRequireDefault(_slice);

var _some = require('./some');

var _some2 = _interopRequireDefault(_some);

var _split = require('./split');

var _split2 = _interopRequireDefault(_split);

var _startsWith = require('./startsWith');

var _startsWith2 = _interopRequireDefault(_startsWith);

var _startsWithAt = require('./startsWithAt');

var _startsWithAt2 = _interopRequireDefault(_startsWithAt);

var _sum = require('./sum');

var _sum2 = _interopRequireDefault(_sum);

var _tail = require('./tail');

var _tail2 = _interopRequireDefault(_tail);

var _take = require('./take');

var _take2 = _interopRequireDefault(_take);

var _takeUntil = require('./takeUntil');

var _takeUntil2 = _interopRequireDefault(_takeUntil);

var _takeWhile = require('./takeWhile');

var _takeWhile2 = _interopRequireDefault(_takeWhile);

var _test = require('./test');

var _test2 = _interopRequireDefault(_test);

var _times = require('./times');

var _times2 = _interopRequireDefault(_times);

var _toLowerCase = require('./toLowerCase');

var _toLowerCase2 = _interopRequireDefault(_toLowerCase);

var _toUpperCase = require('./toUpperCase');

var _toUpperCase2 = _interopRequireDefault(_toUpperCase);

var _trim = require('./trim');

var _trim2 = _interopRequireDefault(_trim);

var _uncurry = require('./uncurry');

var _uncurry2 = _interopRequireDefault(_uncurry);

var _uncurry3 = require('./uncurry3');

var _uncurry32 = _interopRequireDefault(_uncurry3);

var _unfold = require('./unfold');

var _unfold2 = _interopRequireDefault(_unfold);

var _values = require('./values');

var _values2 = _interopRequireDefault(_values);

var _xor = require('./xor');

var _xor2 = _interopRequireDefault(_xor);

exports.always = _always2['default'];
exports.and = _and2['default'];
exports.assign = _assign2['default'];
exports.average = _average2['default'];
exports.between = _between2['default'];
exports.bind = _bind2['default'];
exports.bitAnd = _bitAnd2['default'];
exports.bitOr = _bitOr2['default'];
exports.butLast = _butLast2['default'];
exports.by = _by2['default'];
exports.charAt = _charAt2['default'];
exports.charCodeAt = _charCodeAt2['default'];
exports.codePointAt = _codePointAt2['default'];
exports.compose = _compose2['default'];
exports.composeAll = _composeAll2['default'];
exports.concat = _concat2['default'];
exports.converge = _converge2['default'];
exports.curry = _curry2['default'];
exports.curryRight = _curryRight2['default'];
exports.dec = _dec2['default'];
exports.drop = _drop2['default'];
exports.endsWith = _endsWith2['default'];
exports.endsWithAt = _endsWithAt2['default'];
exports.equal = _equal2['default'];
exports.every = _every2['default'];
exports.exec = _exec2['default'];
exports.explode = _explode2['default'];
exports.filter = _filter2['default'];
exports.flatMap = _flatMap2['default'];
exports.flip = _flip2['default'];
exports.fold = _fold2['default'];
exports.foldRight = _foldRight2['default'];
exports.forEach = _forEach2['default'];
exports.greaterOrEqual = _greaterOrEqual2['default'];
exports.greaterThan = _greaterThan2['default'];
exports.hasOwnProperty = _hasOwnProperty2['default'];
exports.head = _head2['default'];
exports.identity = _identity2['default'];
exports.ifThen = _ifThen2['default'];
exports.ifThenElse = _ifThenElse2['default'];
exports.implode = _implode2['default'];
exports.inc = _inc2['default'];
exports.includes = _includes2['default'];
exports.indexOf = _indexOf2['default'];
exports.isBetween = _isBetween2['default'];
exports.isBoolean = _isBoolean2['default'];
exports.isFalse = _isFalse2['default'];
exports.isFalsy = _isFalsy2['default'];
exports.isFunction = _isFunction2['default'];
exports.isNull = _isNull2['default'];
exports.isNumber = _isNumber2['default'];
exports.isObject = _isObject2['default'];
exports.isPlainObject = _isPlainObject2['default'];
exports.isString = _isString2['default'];
exports.isTrue = _isTrue2['default'];
exports.isTruthy = _isTruthy2['default'];
exports.isTypeOf = _isTypeOf2['default'];
exports.isUndefined = _isUndefined2['default'];
exports.isUnknown = _isUnknown2['default'];
exports.join = _join2['default'];
exports.keys = _keys2['default'];
exports.last = _last2['default'];
exports.lastIndexOf = _lastIndexOf2['default'];
exports.length = _length2['default'];
exports.lessOrEqual = _lessOrEqual2['default'];
exports.lessThan = _lessThan2['default'];
exports.localeCompare = _localeCompare2['default'];
exports.looseEqual = _looseEqual2['default'];
exports.map = _map2['default'];
exports.match = _match2['default'];
exports.max = _max2['default'];
exports.method = _method2['default'];
exports.min = _min2['default'];
exports.minus = _minus2['default'];
exports.nand = _nand2['default'];
exports.noop = _noop2['default'];
exports.nor = _nor2['default'];
exports.normalize = _normalize2['default'];
exports.not = _not2['default'];
exports.nth = _nth2['default'];
exports.omit = _omit2['default'];
exports.or = _or2['default'];
exports.pick = _pick2['default'];
exports.pipe = _pipe2['default'];
exports.pipeAll = _pipeAll2['default'];
exports.plus = _plus2['default'];
exports.product = _product2['default'];
exports.property = _property2['default'];
exports.push = _push2['default'];
exports.put = _put2['default'];
exports.reduce = _reduce2['default'];
exports.reduceRight = _reduceRight2['default'];
exports.repeat = _repeat2['default'];
exports.replace = _replace2['default'];
exports.search = _search2['default'];
exports.shallowClone = _shallowClone2['default'];
exports.shave = _shave2['default'];
exports.signum = _signum2['default'];
exports.slice = _slice2['default'];
exports.some = _some2['default'];
exports.split = _split2['default'];
exports.startsWith = _startsWith2['default'];
exports.startsWithAt = _startsWithAt2['default'];
exports.sum = _sum2['default'];
exports.tail = _tail2['default'];
exports.take = _take2['default'];
exports.takeUntil = _takeUntil2['default'];
exports.takeWhile = _takeWhile2['default'];
exports.test = _test2['default'];
exports.times = _times2['default'];
exports.toLowerCase = _toLowerCase2['default'];
exports.toUpperCase = _toUpperCase2['default'];
exports.trim = _trim2['default'];
exports.uncurry = _uncurry2['default'];
exports.uncurry3 = _uncurry32['default'];
exports.unfold = _unfold2['default'];
exports.values = _values2['default'];
exports.xor = _xor2['default'];
